python/pyspark/sql/dataframe.py

```diff
@@ -1721,7 +1721,8 @@ def toPandas(self):
         1    5    Bob
         """
         import pandas as pd
-        return pd.DataFrame.from_records(self.collect(), columns=self.columns)
+        dtype = {field.name: _to_numpy_type(field.dataType) for field in self.schema}
+        return pd.DataFrame.from_records(self.collect(), columns=self.columns).astype(dtype)
```

Comment:
This is probably the easiest way to assign the types, but data is still loaded and inferred then the `astype` will then cast the data and I'm not sure if it will make a pass over the data or do it lazily.  A more ideal way would be to not use `from_records` but then I think the data would need to be broken up into columns.